### PR TITLE
Grid documentation update

### PR DESCRIPTION
**Pull request by rkaraivanov**

### CODE REVIEW

File: en/components/grid.md [hunk 1]
```diff
@@ -22,345 +22,503 @@ _keywords: Ignite UI for Angular, UI controls, Angular widgets, web widgets, UI
...
+    public grid: IgxGridComponent;
 
 ### Usage
 
 Now that we have the grid module imported, let’s get started with a basic configuration of the **igx-grid** that binds to local data:
 
-```html
-<igx-grid #grid1 id="grid1" [data]="localData" [autoGenerate]="true"></igx-grid>
-```
+    <igx-grid #grid1 id="grid1" [data]="localData" [autoGenerate]="true"></igx-grid>
```
Comment (ig-ealbert) [inline en/components/grid.md]: Is there a reason that the code delimiters were removed?

File: en/components/grid.md [hunk 2]
```diff
@@ -22,345 +22,503 @@ _keywords: Ignite UI for Angular, UI controls, Angular widgets, web widgets, UI
...
 
-Each of the components, directives and helper classes in the _IgxGridModule_ can be imported either through the _grid.component_ or through
-the main bundle in _igniteui-angular_. While you don't need to import all of them to instantiate and use the grid, you usually will import them
-(or your editor will auto-import them for you) when declaring types that are part of the grid API.
+Each of the components, directives and helper classes in the _IgxGridModule_ can be imported either through the _grid_ sub-package or through the main bundle in _igniteui-angular_. While you don't need to import all of them to instantiate and use the grid, you usually will import them (or your editor will auto-import them for you) when declaring types that are part of the grid API.
 
-```typescript
-import { IgxGridComponent } from 'igniteui-angular/grid/grid.component';
-...
+    import { IgxGridComponent } from 'igniteui-angular/grid/';
```
Comment (ig-ealbert) [inline en/components/grid.md]: Why were the code delimiters removed?  In this way, you won't get syntax highlighting, right?

File: en/components/grid.md [hunk 3]
```diff
@@ -22,345 +22,503 @@ _keywords: Ignite UI for Angular, UI controls, Angular widgets, web widgets, UI
...
-```html
-<igx-grid #grid1 [data]="data | async" [autoGenerate]="false" [paging]="true" [perPage]="6" (onColumnInit)="initColumns($event)"
-    (onCellSelection)="selectCell($event)">
-    <igx-column field="Name" [sortable]="true" header=" " [filtering]="true"></igx-column>
-    <igx-column field="AthleteNumber" [sortable]="true" header="Athlete number"></igx-column>
-    <igx-column field="TrackProgress" header="Track progress">
-        <ng-template igxCell let-col="column" let-ri="rowIndex" let-item="item">
-            <igx-linear-bar [striped]="false" [value]="item" [max]="100">
-            </igx-linear-bar>
+    <igx-grid #grid1 [data]="data | async" [autoGenerate]="false" [paging]="true" [perPage]="6" (onColumnInit)="initColumns($event)"
```
Comment (ig-ealbert) [inline en/components/grid.md]: Let's just assume I have the same comment for all code snippets that are not delimited.

File: en/components/grid.md [hunk 4]
```diff
@@ -22,345 +22,503 @@ _keywords: Ignite UI for Angular, UI controls, Angular widgets, web widgets, UI
...
-      column.filtering = true;
-      column.sortable = true;
-      column.editable = true;
-    }
-}
-```
+Make sure to check the API for the [IgxGridCellComponent](#igxgridcellcomponent) in order to get accustomed with the provided properties you can use in your templates.
 
-The code above will make the **ProductName** column sortable, filterable, and editable and will instantiate the corresponding features UI (like inputs for editing, save dialogs, etc.).
+Each of the column templates can be changed programmatically at any point through the column object itself. For example in the code below, we have declared two templates for our user data. In out Typescript code we'll get references to the templates itself and then based on some condition we will render the appropriate for the column in our application.
```
Comment (ig-ealbert) [inline en/components/grid.md]: In *our* Typescript code we'll get references to the template *themselves*
-or-
... to the template itself (if there is only one)

...we will render the appropriate [word is missing here] for the column.  Render the appropriate template?

File: en/components/grid.md [hunk 5]
```diff
@@ -22,345 +22,503 @@ _keywords: Ignite UI for Angular, UI controls, Angular widgets, web widgets, UI
...
+            HttpClientModule
+            ...
+        ],
+        providers: [
+            NorthwindService
+        ]
+    })
+    export class AppModule {}
+
+After implementing the service we will inject it in our components constructor and use it to retrieve the data. The `ngOnInit` lifecycle hook is a good place to dispatch the initial request.
```
Comment (ig-ealbert) [inline en/components/grid.md]: component's constructor.

File: en/components/grid.md [hunk 6]
```diff
@@ -22,345 +22,503 @@ _keywords: Ignite UI for Angular, UI controls, Angular widgets, web widgets, UI
...
 
-public deleteRow(event) {
-    this.selectedRow = Object.assign({}, this.grid1.getRow(this.selectedCell.rowIndex));
-    this.grid1.deleteRow(this.selectedCell.rowIndex);
-    this.selectedCell = {};
-    this.snax.message = `Row with ID ${this.selectedRow.record.ID} was deleted`;
-    this.snax.show();
-}
-```
+The grid component exposes and `addRow` method which will add the provided data to the data source itself.
```
Comment (ig-ealbert) [inline en/components/grid.md]: `and` should be either `an` or `the`.

File: en/components/grid.md [hunk 7]
```diff
@@ -22,345 +22,503 @@ _keywords: Ignite UI for Angular, UI controls, Angular widgets, web widgets, UI
...
+        // Updating the whole row
+        this.grid.updateRow(newData, this.selectedCell.rowIndex);
+
+        // Just a particualr cell through the Grid API
+        this.grid.updateCell(newData, this.selectedCell.rowIndex, this.selectedCell.column.field);
+
+        // Directly using the cell `update` method
+        this.selectedCell.update(newData);
+
+#### Deleteing data from the grid
```
Comment (ig-ealbert) [inline en/components/grid.md]: Deleting

File: en/components/grid.md [hunk 8]
```diff
@@ -22,345 +22,503 @@ _keywords: Ignite UI for Angular, UI controls, Angular widgets, web widgets, UI
...
+You can filter any column or a combination of columns through the grid API. The grid exposes two methods for this task:
+
+*   `filter` - filter a single column or a combination of columns.
+
+        // Single column filtering
+
+        // Filter the `ProductName` column for values which `contains` the `myproduct` substring, ignoring case
+        this.grid.filter('ProductName', 'myproduct', STRING_FILTERS.contains, true);
+
+The only required parameters are the column field and the filtering term. Both the condition and the case sensitivity will be infered from the column properties if not provided. In the case of multiple filtering, the method accepts an array of filtering expressions.
```
Comment (ig-ealbert) [inline en/components/grid.md]: I think "inferred" has 2 rs but I'm not sure. :)

File: en/components/grid.md [hunk 9]
```diff
@@ -22,345 +22,503 @@ _keywords: Ignite UI for Angular, UI controls, Angular widgets, web widgets, UI
...
+
+        // Removes the sorting state from the ProductName column
+        this.grid.clearSort('ProductName');
+
+        // Removes the sorting state from every column in the grid
+        this.grid.clearSort();
+
+#### Initial sorting state
+
+It is possible to set the initial sorting state of the grid by an array of expressions to the `sortingExpressions` property of the grid.
```
Comment (ig-ealbert) [inline en/components/grid.md]: by setting an array?

File: en/components/grid.md [hunk 10]
```diff
@@ -22,345 +22,503 @@ _keywords: Ignite UI for Angular, UI controls, Angular widgets, web widgets, UI
...
-| `onBeforeProcess` | Emit binding behavior                                                       |
+|Name|Description|
+|--- |--- |
+|_Event emitters_|_Notify for a change_|
+|`onEditDone`|Emitted when a cell value changes. Returns `{ currentValue: any, newValue: any }`|
+|`onSelection`|Emitted when a cell is selected. Returns the cell object.|
+|`onColumnInit`|Emitted when the grid columns are initialized. Returns the column object.|
+|`onSortingDone`|Emitted when sorting is performed through the UI. Returns the sorting expression.|
+|`onFilteringDone`|Emitted when filtering is performed through the UI. Returns the filtering expression.|
+|`onPagingDone`|Emitted when paging is performed. Returns an object consisiting of the previous and the new page.|
```
Comment (ig-ealbert) [inline en/components/grid.md]: consisting

File: en/components/grid.md [hunk 11]
```diff
@@ -377,40 +535,178 @@ Defining handlers for these event emitters is done using declarative event bindi
...
-| `updateRow`                                      | Update record from the grid data container               |
-| `updateCell`                                     | Update grid cell by index, column field and passed value |
-| `sortColumn`                                     | Sort grid column                                         |
-| `paginate`                                       | Change the current page by passed number                 |
+|Signature|Description|
+|--- |--- |
+|`getColumnByName(name: string)`|Returns the column object with field property equal to `name` or `undefined` if no such column exists.|
+|`getCellByColumn(rowIndex: number, columnField: string)`|Returns the cell object in column with `columnField` and row with `rowIndex` or `undefined`.|
+|`addRow(data: any)`|Creates a new row object and adds the `data` record to the end of the data source.|
+|`deleteRow(rowIndex: number)`|Removes the row object and the data recrod corresponding to it from the data source.|
```
Comment (ig-ealbert) [inline en/components/grid.md]: and the corresponding data record from the data source.

File: en/components/grid.md [hunk 12]
```diff
@@ -377,40 +535,178 @@ Defining handlers for these event emitters is done using declarative event bindi
...
+|`filter(column: string, value: any, condition?, ignoreCase?: boolean)`|Filters a single column. Check the available [filtering conditions](#filtering-conditions)|
+|`filter(expressions: Array)`|Filters the grid columns based on the provided array of filtering expressions.|
+|`filterGlobal(value: any, condition? ignoreCase?)`|Filters all the columns in the grid.|
+|`clearFilter(name?: string)`|If `name` is provided, clears the filtering state of the corresponding column, otherwise clears the filtering state of all columns.|
+|`sort(name: string, direction, ignorecase)`|Sorts a single column.|
+|`sort(expressions: Array)`|Sorts the grid columns based on the provided array of sorting expressions.|
+|`clearSort(name?: string)`|If `name` is provided, clears the sorting state of the corresponding column, otherwise clears the sorting state of all columns.|
+|`previousPage()`|Goes to the previous page if paging is enabled and the current page is not the first.|
+|`nextPage()`|Goes to the next page if paging is enabled and current page is not the last.|
+|`paginate(page: number)`|Goes to the specified page if paging is enabled. Page indexes are 0 based.|
```
Comment (ig-ealbert) [inline en/components/grid.md]: Page indices?  I think the plural of "index" is "indices" but this is minor.  It's understandable.

File: en/components/grid.md [hunk 13]
```diff
@@ -377,40 +535,178 @@ Defining handlers for these event emitters is done using declarative event bindi
...
+|`markForCheck()`|Manually triggers a changed detection cycle of the grid and its children.|
```
Comment (ig-ealbert) [inline en/components/grid.md]: change detection?  Or does it trigger a detection cycle that has changed already?

File: en/components/grid.md [hunk 14]
```diff
@@ -377,40 +535,178 @@ Defining handlers for these event emitters is done using declarative event bindi
...
+<div class="divider--half"></div>
+
+### Getters/Setters
+
+|Name|Type|Getter|Setter|Description|
+|--- |--- |--- |--- |--- |
+|`bodyTemplate`|TemplateRef|Yes|Yes|Get/Set a reference to a template which will be applied to the cells in the column.|
+|`headerTemplate`|TemplateRef|Yes|Yes|Get/Set a reference to a template which will be applied to the column header.|
+|`footerTemplate`|TemplateRef|Yes|Yes|Get/Set a reference to a template which will be applied to the column footer.|
+|`inlineEditorTemplate`|TemplateRef|Yes|Yes|Get/Set a reference to a template which will be applied as a cell enters in edit mode.|
```
Comment (ig-ealbert) [inline en/components/grid.md]: enters edit mode.

File: en/components/grid.md [hunk 15]
```diff
@@ -377,40 +535,178 @@ Defining handlers for these event emitters is done using declarative event bindi
...
+|`empty`|`(target: any)`|Returns true if `target` is either `null`, `undefined` or `NaN`.|
+|`notEmpty`|`(target: any)`|Returns true if `target` is not `null`, `undefined` or `NaN`.|
+
+<div class="divider--half"></div>
+
+### Boolean types
+
+|Name|Signature|Description|
+|--- |--- |--- |
+|`true`|`(target: boolean)`|Returns if target is truthy.|
```
Comment (ig-ealbert) [inline en/components/grid.md]: Should `target` here be marked with backticks fro consistency?  Same on line 651.

File: en/components/grid.md (hunk 3 above)
Comment (kdinev) [reply en/components/grid.md]: @ig-ealbert The code is indented as code, so it will still show as such.

**Comment by kdinev** [comment]
@ig-ealbert Indentation is still OK for code, but I agree with you that syntax is better than indentation because it's less error prone for when somebody else performs changes to the documents.

File: en/components/grid.md [hunk 16]
```diff
@@ -61,11 +61,9 @@ Now that we have the grid module imported, let’s get started with a basic conf
 <igx-grid #grid1 id="grid1" [data]="localData" [autoGenerate]="true"></igx-grid>
 ```
 
-The **id** property is a string value and is the unique identifier of the grid, while **data** binds the grid, in this case to local data.
-
-The **autoGenerate** property tells the **igx-grid** to autogenerate columns based on the data source fields. Otherwise, the developer needs to explicitly define the columns and the mapping to the data source fields.
+The **id** property is a string value and is the unique identifier of the grid which will be autogenerated if not provided, while **data** binds the grid, in this case to local data.
 
-<div class="divider--half"></div>
+The **autoGenerate** property tells the **igx-grid** to autogenerate columns based on the data source fields. It will also try to deduce the appropriate data type for the column if possible. Otherwise, the developer needs to explicitly define the columns and the mapping to the data source fields.
```
Comment (kdinev) [inline en/components/grid.md]: Refer to columns as column objects and their type. "to auto generate the grid's `IgxColumnComponent` objects...`.

File: en/components/grid.md [hunk 17]
```diff
@@ -75,39 +73,127 @@ Let's turn the **autoGenerate** property off and define the columns collection i
...
+    <igx-column dataType="boolean" field="Subscribtion">
+        <ng-template igxCell let-cell="cell">
+            <!-- Bind the cell value through the ngModel directive and update the data source when the value is changed in the template -->
+            <input type="checkbox" [ngModel]="cell.value" (ngModelChange)="cell.update($event)" />
+        </ng-template>
+    </igx-column>
+<igx-grid>
+```
+
+The column also accepts one last template that will be used when a cell is in edit mode. As with the other column templates, the provided context object is again the cell value and the cell object itself. Of course in order to have your template show up make sure to set the `editable` property of the column to true.
```
Comment (kdinev) [inline en/components/grid.md]: Of course, in order to make the edit-mode template accessible to end users, you need to set the `editable` property of the `IgxColumnComponent` to `true`.

File: en/components/grid.md [hunk 18]
```diff
@@ -75,39 +73,127 @@ Let's turn the **autoGenerate** property off and define the columns collection i
...
+            Enter the new price tag
+        </label>
+        <input name="price" type="number" [ngModel]="cell.value" (ngModelChange)="cell.update(convertToNumber($event))" />
+    </ng-template>
+</igx-column>
+```
+
+Make sure to check the API for the [IgxGridCellComponent](#igxgridcellcomponent) in order to get accustomed with the provided properties you can use in your templates.
+
+Each of the column templates can be changed programmatically at any point through the column object itself. For example in the code below, we have declared two templates for our user data. In our Typescript code we'll get references to the templates themselves and then based on some condition we will render the appropriate template for the column in our application.
```
Comment (kdinev) [inline en/components/grid.md]: "... through the `IgxColumnComponent` object itself."

Also refer to Typescript as TypeScript.

File: en/components/grid.md [hunk 19]
```diff
@@ -75,39 +73,127 @@ Let's turn the **autoGenerate** property off and define the columns collection i
...
+        column.sortable = true;
+        column.editable = true;
     }
 }
 ```
 
-The code above will make the **ProductName** column sortable, filterable, and editable and will instantiate the corresponding features UI (like inputs for editing, save dialogs, etc.).
-
-<div class="divider--half"></div>
+The code above will make the **ProductName** column sortable, filterable, and editable and will instantiate the corresponding features UI (like inputs for editing, etc.).
```
Comment (kdinev) [inline en/components/grid.md]: The code above will make the **ProductName** column **filterable**, sortable, and editable, and will... (keep the order the same as in the code example).

File: en/components/grid.md [hunk 20]
```diff
@@ -75,39 +73,127 @@ Let's turn the **autoGenerate** property off and define the columns collection i
...
 
 ### Data binding
 
-Before going any further with the grid we want to change the grid to bind to remote data service, which is a common scenario in real production applications. A good practice is to separate all data fetching related logic in a separate data service, so we are going to create a service which
-will handle the fetching of data from the server.
+Before going any further with the grid we want to change the grid to bind to remote data service, which is a common scenario in real production applications. A good practice is to separate all data fetching related logic in a separate data service, so we are going to create a service which will handle the fetching of data from the server.
```
Comment (kdinev) [inline en/components/grid.md]: Don't use language such as "real production applications" - this is internal language. Use something like - "large-scale applications", "enterprise-grade applications".

File: en/components/grid.md [hunk 21]
```diff
@@ -121,14 +207,11 @@ import { of } from 'rxjs/observable/of';
 import { catchError, map } from 'rxjs/operators';
 ```
 
-We're importing the `Injectable` decorator which is an [essential ingredient](https://angular.io/guide/dependency-injection) in every Angular service definition. The `HttpClient` will provide us with the functionality to communicate with
-backend services. It returns an `Observable` of some result on which we will subscribe in our grid component.
+We're importing the `Injectable` decorator which is an [essential ingredient](https://angular.io/guide/dependency-injection) in every Angular service definition. The `HttpClient` will provide us with the functionality to communicate with backend services. It returns an `Observable` of some result on which we will subscribe in our grid component.
```
Comment (kdinev) [inline en/components/grid.md]: "**to** which we will subscribe..."

File: en/components/grid.md [hunk 22]
```diff
@@ -234,141 +310,289 @@ export class MyComponent implements OnInit {
...
-**Note**: The grid `autoGenerate` property is best to be avoided when binding to remote data for now. It assumes that the data is available in order
-to inspect it and generate the appropriate columns. This is usually not the case until the remote service responds, and
-the grid will throw an error. Making `autoGenerate` available, when binding to remote service, is on our roadmap for future versions.
+**Note**: The grid `autoGenerate` property is best to be avoided when binding to remote data for now. It assumes that the data is available in order to inspect it and generate the appropriate columns. This is usually not the case until the remote service responds, and the grid will throw an error. Making `autoGenerate` available, when binding to remote service, is on our roadmap for future versions.
 
-<div class="divider--half"></div>
+### Virtualization
 
-### CRUD operations though the API
+The grid supports both horizontal and vertical virtualization. In order to activate it make sure to set either the `height` or `width` property of the grid.
```
Comment (kdinev) [inline en/components/grid.md]: DOM virtualization

File: en/components/grid.md [hunk 23]
```diff
@@ -234,141 +310,289 @@ export class MyComponent implements OnInit {
...
 ```html
 <button igxButton igxRipple (click)="deleteRow($event)">Delete Row</button>
 ```
 
-<div class="divider--half"></div>
-
 ### Paging
 
-**Paging** is initialized on the root **igx-grid** component, and is configurable via the `paging` and `perPage` inputs. Paging is a Boolean property that controls whether the feature is enabled and the perPage property controls the visible records per page. Let’s update our grid to enable paging:
+**Paging** is initialized on the root **igx-grid** component, and is configurable via the `paging` and `perPage` inputs. Paging is a Boolean property that controls whether the feature is enabled and the `perPage` property controls the visible records per page. Let’s update our grid to enable paging:
```
Comment (kdinev) [inline en/components/grid.md]: Note that the `perPage` input has to be a number larger than the virtualization chunk size.

File: en/components/grid.md [hunk 24]
```diff
@@ -540,7 +607,7 @@ Here is a list of all public methods exposed by the **igx-grid**:
 |`getColumnByName(name: string)`|Returns the column object with field property equal to `name` or `undefined` if no such column exists.|
 |`getCellByColumn(rowIndex: number, columnField: string)`|Returns the cell object in column with `columnField` and row with `rowIndex` or `undefined`.|
 |`addRow(data: any)`|Creates a new row object and adds the `data` record to the end of the data source.|
-|`deleteRow(rowIndex: number)`|Removes the row object and the data recrod corresponding to it from the data source.|
+|`deleteRow(rowIndex: number)`|Removes the row object and and the corresponding data record from the data source.|
```
Comment (ig-ealbert) [inline en/components/grid.md]: There are two instances of "and" here.

File: en/components/grid.md [hunk 25]
```diff
@@ -234,141 +311,289 @@ export class MyComponent implements OnInit {
...
+|`autoGenerate`|boolean|Autogenerate grid's columns, default value is _false_|
+|`paging`|bool|Enables the paging feature. Defaults to _false_.|
+|`perPage`|number|Visible items per page, default is 15|
+|`filteringLogic`|FilteringLogic|The filtering logic of the grid. Defaults to _AND_.|
+|`filteringExpressions`|Array|The filtering state of the grid.|
+|`sortingExpressions`|Array|The sorting state of the grid.|
+|`height`|string|The height of the grid element. You can pass values such as `1000px`, `75%`, etc.|
+|`width`|string|The width of the grid element. You can pass values such as `1000px`, `75%`, etc.|
+|`evenRowCSS`|string|Additional styling classes applied to all even rows in the grid.|
+|`oddRowCSS`|string|Additional styling classses applied to all odd rows in the grid.|
```
Comment (ig-ealbert) [inline en/components/grid.md]: One too many "s" in "classes".

File: en/components/grid.md [hunk 26]
```diff
@@ -377,40 +602,180 @@ Defining handlers for these event emitters is done using declarative event bindi
...
+|`header`|string|Column header text|
+|`sortable`|boolean|Set column to be sorted or not|
+|`editable`|boolean|Set column values to be editable|
+|`filterable`|boolean|Set column values to be filterable|
+|`hidden`|boolean|Visibility of the column|
+|`movable`|boolean|Column moving|
+|`width`|string|Columns width|
+|`headerClasses`|string|Additional CSS classes applied to the header element.|
+|`cellClasses`|string|Additional CSS classes applied to the cells in this column.|
+|`formatter`|Function|A function used to "template" the values of the cells without the need to pass a cell template the column.|
```
Comment (ig-ealbert) [inline en/components/grid.md]: pass a cell template to the column.

File: en/components/grid.md [hunk 27]
```diff
@@ -377,40 +602,180 @@ Defining handlers for these event emitters is done using declarative event bindi
...
+|`empty`|`(target: any)`|Returns true if `target` is either `null`, `undefined` or `NaN`.|
+|`notEmpty`|`(target: any)`|Returns true if `target` is not `null`, `undefined` or `NaN`.|
+
+<div class="divider--half"></div>
+
+### Boolean types
+
+|Name|Signature|Description|
+|--- |--- |--- |
+|`true`|`(target: boolean)`|Returns if `target` is truthy.|
```
Comment (ig-ealbert) [inline en/components/grid.md]: Returns true?

**Pull request merged**